ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+    def self.execute_activity(activity, *args, **left_off_for_brevity)
+    def self.execute_local_activity(activity, *args, **left_off_for_brevity)
+
+    # Returns ChildWorkflowHandle.
+    def self.start_child_workflow(workflow, *args, **left_off_for_brevity)
+    # Returns actual result.
+    def self.execute_child_workflow(workflow, *args, **left_off_for_brevity)
+
+    # These are a custom hash class that only supports []= mutation and
+    # validates input as a signal/query/update definition. Basically they will
+    # be simple delegators to frozen hashes only updatable via index. Symbol
+    # keys or nil key for dynamic. All values are "definition" objects.
+    def self.signal_handlers
+    def self.query_handlers
+    def self.update_handlers
+
+    def info
+    def current_update_info
+
+    def self.search_attributes
+    def self.upsert_search_attributes(*updates)
+    def self.raw_memo
+    def self.memo_value(name)
+    def self.upsert_memo(*updates)
+
+    def self.external_workflow_handle(id, run_id: nil)
+
+    def self.patched(id)
+    def self.deprecate_patch(id)
+
+    def self.cancellation
+    # Just a shortcut to getting a future that fails on cancel. Good for
+    # Future.any_of use.
+    def self.cancellation_future
+    def self.logger
+    def self.now
+    def self.payload_converter
+    def self.random
+    def self.uuid
+    def self.wait_condition(&)
+
+    class ContinueAsNewError < Error
+      def initialize(*args, workflow: nil, **left_off_for_brevity)
+    end
+
+    class Unsafe
+      def replaying?
+      # See TracePoint later on why this is present.
+      def illegal_call_tracing_disabled(&)
+    end
+
+    # Other nested classes omitted for brevity
+  end
+end
+```
+
+Notes:
+
+* Notice lack of `timeout` on `wait_condition`, we will support Ruby `Timeout` via the fiber scheduler.
+* Notice lack of `sleep`, we will support Ruby `sleep` via the fiber scheduler.
+* A `cancellation` can be provided to every async thing, but by default it's the workflow-level one.
+* A workflow builds a `Workflow::Definition` object which contains metadata about the structure including a set of
+  `Workflow::Definition::Signal`s, `Workflow::Definition::Query`s, and `Workflow::Definition::Update`s built from the
+  handlers (but of course handlers can be added/removed at runtime via the `_handlers` hashes).
+* The `workflow_` methods above can only be called during class definition time not workflow runtime.
+* The runtime class methods above can only be called during runtime and can only be called on the `Workflow` class
+  itself. This means `Workflow.wait_condition` is ok, but `FooWorkflow.wait_condition` or `self.class.wait_condition`
+  is not.
+  * ❓ Is this acceptable? This is a consequence of us wanting to share workflow context class with workflow base
+    class. Basically at the top of these calls there's a `if self != Workflow` check.
```

Comment:
That is definitely a good question and subject to debate and is the reason I put "`Temporalio::Workflow` is both the class to extend and the class with class methods for calling statically to do things." at the top of the controversial decisions.

So basically you have these options:

### Option 1 - Together

```ruby
class MyWorkflow < Temporalio::Workflow
  def execute
    Temporalio::Workflow.wait_condition { false }
  end
end
```

### Option 2 - Context class named separate

```ruby
class MyWorkflow < Temporalio::Workflow
  def execute
    Temporalio::WorkflowContext.wait_condition { false }
  end
end
```

### Option 3 - Workflow class named separate

```ruby
class MyWorkflow < Temporalio::WorkflowImpl
  def execute
    Temporalio::Workflow.wait_condition { false }
  end
end
```

I am using `WorkflowImpl` here because there will be a `Temporalio::Workflow::Definition` that represents the static definition built from the impl (and signal, query, and definition classes, same as there is already `Temporalio::Activity::Definition`). Open to a better name, I just struggled to come up with one.

### Option 4 - Don't even require class extension

```ruby
class MyWorkflow
  def execute
    Temporalio::Workflow.wait_condition { false }
  end
end
```

You aren't buying _that_ much with a class extension, but it is clear and is common in Ruby and is what we do for activities.

Thoughts/preference? Any other/better options here? Mixins would suffer the same class-for-two-purposes reuse issue.

Reply:
I like option 3. Clear separation, just as terse as option 1.

Reply:
Yeah, which is basically the same as option 2 just different on who gets the "Workflow" part. I will say I like "Context" for users more than I like "WorkflowImpl" for extenders, but I do like callers of workflow utilities to only need `Temporalio::Workflow` instead of `Temporalio::Workflow::Context` (and what drove my original approach). Either way, I am starting to lean towards separating what you extend from what you invoke. Any ideas for other names though besides either "Context" for workflow utilities or "WorkflowImpl" for the extension class?

Reply:
Not really lol. Those are pretty obvious choices. Maybe `WorkflowDefinition` like I suggested instead of impl. I would err on making that longer since you type it way less.

Reply:
Ok, after some internal discussion, I think I'm landing at something like:

```ruby
module Temporalio
  class Workflow
    class << self
      protected

      def workflow_name(name)
        raise 'TODO'
      end
    end

    module API
      def execute_activity(activity, blah1, blah2)
        raise 'TODO'
      end
    end

    extend API
  end
end
```

So basically you have protected class methods for adjusting the workflow definition, then an `API` set of module methods in the workflow class by default (`extend` in this way I think avoids YARD assuming they are part of the class, still investigating). This will allow others to write utilities for workflows using `Temporalio::Workflow::API` directly (or extending it themselves too). Doing a bit more investigation before accepting this approach.

Reply:
After discussion, proposal updated to require workflows extend from `Temporalio::Workflow::Definition` and use module functions on `Temporalio::Workflow`. As a consistency consequence, activities will now also have to extend from `Temporalio::Activity::Definition`.